crates/sui-core/src/authority/consensus_tx_status_cache.rs

```diff
@@ -0,0 +1,316 @@
+
+impl ConsensusTxStatusCache {
+    pub fn new(gc_depth: u64) -> Self {
+        Self {
+            gc_depth,
+            ..Default::default()
+        }
+    }
+
+    pub fn set_transaction_status(
+        &self,
+        transaction_position: ConsensusTxPosition,
+        status: ConsensusTxStatus,
+    ) {
+        let mut inner = self.inner.write();
+        if let Some(last_committed_round) = inner.last_committed_round {
+            if transaction_position.block.round as u64 + self.gc_depth < last_committed_round {
+                return;
+            }
+        }
+        let old_status = inner
+            .transaction_status
+            .insert(transaction_position, status);
+        if old_status.is_none() {
+            inner
+                .round_lookup_map
+                .entry(transaction_position.block.round as u64)
+                .or_default()
+                .insert(transaction_position);
+        }
+        self.status_notify_read
+            .notify(&transaction_position, &status);
+    }
+
+    pub async fn notify_read_transaction_status(
+        &self,
+        transaction_position: ConsensusTxPosition,
+        old_status: Option<ConsensusTxStatus>,
+    ) -> NotifyReadConsensusTxStatusResult {
+        let registration = self.status_notify_read.register_one(&transaction_position);
+        {
+            let inner = self.inner.read();
+            if let Some(status) = inner.transaction_status.get(&transaction_position) {
+                if Some(status) != old_status.as_ref() {
+                    if let Some(old_status) = old_status {
+                        // The only scenario where the status may change, is when the transaction
+                        // is initially fastpath certified, and then later finalized or rejected.
+                        assert_eq!(old_status, ConsensusTxStatus::FastpathCertified);
+                    }
+                    return NotifyReadConsensusTxStatusResult::Status(*status);
+                }
+            }
+            // Inner read lock dropped here.
+        }
+
+        let expiration_check = async {
+            loop {
+                {
+                    let inner = self.inner.read();
+                    if let Some(last_committed_round) = inner.last_committed_round {
+                        if transaction_position.block.round as u64 + self.gc_depth
+                            < last_committed_round
+                        {
+                            return last_committed_round;
+                        }
+                    }
+                }
+                sleep(Duration::from_millis(50)).await;
+            }
+        };
+        tokio::select! {
+            status = registration => NotifyReadConsensusTxStatusResult::Status(status),
+            last_committed_round = expiration_check => NotifyReadConsensusTxStatusResult::Expired(last_committed_round),
+        }
+    }
+
+    pub async fn update_last_committed_round(&self, round: u64) {
```

Comment:
minor: maybe s/last_committed_round/last_committed_leader_round in this file and callsites to be more precise. In future, committed leader block will actually not be part of the Commit.